tokio-net/src/driver/reactor/sharded_slab/page/mod.rs

```diff
@@ -0,0 +1,299 @@
+/// A page address encodes the location of a slot within a shard (the page
+/// number and offset within that page) as a single linear value.
+#[repr(transparent)]
+#[derive(Copy, Clone, Eq, PartialEq, PartialOrd, Ord)]
+pub(crate) struct Addr {
+    addr: usize,
+}
+
+impl Addr {
+    const NULL: usize = Self::BITS + 1;
+    const INDEX_SHIFT: usize = INITIAL_PAGE_SIZE.trailing_zeros() as usize + 1;
+
+    pub(crate) fn index(self) -> usize {
+        // Since every page is twice as large as the previous page, and all page sizes
+        // are powers of two, we can determine the page index that contains a given
+        // address by shifting the address down by the smallest page size and
+        // looking at how many twos places necessary to represent that number,
+        // telling us what power of two page size it fits inside of. We can
+        // determine the number of twos places by counting the number of leading
+        // zeros (unused twos places) in the number's binary representation, and
+        // subtracting that count from the total number of bits in a word.
+        WIDTH - ((self.addr + INITIAL_PAGE_SIZE) >> Self::INDEX_SHIFT).leading_zeros() as usize
+    }
+
+    pub(crate) fn offset(self) -> usize {
+        self.addr
+    }
+}
+
+pub(super) fn size(n: usize) -> usize {
+    INITIAL_PAGE_SIZE * 2usize.pow(n as _)
+}
+
+impl Pack for Addr {
+    const LEN: usize = super::MAX_PAGES + Self::INDEX_SHIFT;
+
+    type Prev = ();
+
+    fn as_usize(&self) -> usize {
+        self.addr
+    }
+
+    fn from_usize(addr: usize) -> Self {
+        debug_assert!(addr <= Self::BITS);
+        Self { addr }
+    }
+}
+
+pub(in crate::driver) type Iter<'a> =
+    std::iter::FilterMap<std::slice::Iter<'a, Slot>, fn(&'a Slot) -> Option<&'a ScheduledIo>>;
+
+pub(crate) struct Local {
+    head: CausalCell<usize>,
+}
+
+pub(crate) struct Shared {
+    remote_head: AtomicUsize,
+    size: usize,
+    prev_sz: usize,
+    slab: CausalCell<Option<Box<[Slot]>>>,
+}
+
+impl Local {
+    pub(crate) fn new() -> Self {
+        Self {
+            head: CausalCell::new(0),
+        }
+    }
+
+    #[inline(always)]
+    fn head(&self) -> usize {
+        self.head.with(|head| unsafe { *head })
+    }
+
+    #[inline(always)]
+    fn set_head(&self, new_head: usize) {
+        self.head.with_mut(|head| unsafe {
+            *head = new_head;
+        })
+    }
+}
+
+impl Shared {
+    const NULL: usize = Addr::NULL;
+
+    pub(crate) fn new(size: usize, prev_sz: usize) -> Self {
+        Self {
+            prev_sz,
+            size,
+            remote_head: AtomicUsize::new(Self::NULL),
+            slab: CausalCell::new(None),
+        }
+    }
+
+    #[cold]
+    fn fill(&self) {
+        #[cfg(test)]
+        test_println!("-> alloc new page ({})", self.size);
+
+        debug_assert!(self.slab.with(|s| unsafe { (*s).is_none() }));
+
+        let mut slab = Vec::with_capacity(self.size);
+        slab.extend((1..self.size).map(Slot::new));
```

Comment:
We don't want to eagerly initialize slots. This could be very expensive. We want to amortize the cost it by doing it lazily.

Reply:
@carllerche we need to _allocate_ the whole page eagerly (rather than using `Vec::push`) to avoid reallocating while data is being accessed by another thread. Do you mean that we could allocate uninitialized memory and only initialize the slots as needed? Because (AFAICT) that calls for an additional atomic tracking the last written slot.

Reply:
actually...it occurs to me just now that we _don't_ need to use `MaybeUninit` and a counter tracking the last initialized slot. We can just add new slots with `Vec::push`, and use `Vec::with_capacity` to avoid ever triggering a realloc in `push`. That way, we initialize the slots lazily but have a single allocation that we never need to reallocate. 

Since pushes would only occur in `insert`, which is called only from the local thread (or when the slab is locked in the single-shard case), we don't need to use an atomic.

Reply:
@carllerche The approach I mentioned above _does_ appear to work, but it makes it much harder to test the code using `loom` (the `Vec` mutation can occur concurrently with accesses which we _know_ are to slots that will not be mutated, but loom doesn't know that). Using `CausalCell<MaybeUninit<Slot>>`plays much nicer with `loom`, as the mutation can be tracked at the level of each slot in the slab. 

I have a branch that makes that change: https://github.com/tokio-rs/tokio/compare/eliza/net-driver-slab...eliza/net-driver-slab-lazy-init?expand=1#diff-398083c85d9c10f01b1c6aeb84bcdaf1 However, I'm not sure if this is worth it; as far as I can tell, this requires adding an atomic tracking the last slot that was initialized, and `insert`, `get`, and `remove` operations all have to access that atomic. I'd be surprised if lazily initializing slots makes a big enough performance difference to be worth adding so much more unsafe code. What do you think?